efi/shim.go

```diff
@@ -0,0 +1,366 @@
+type shimVersion struct {
+	Major uint
+	Minor uint
+}
+
+// parseShimVersion returns a new shimVersion from the supplied string
+func parseShimVersion(version string) (shimVersion, error) {
+	m := shimVersionRE.FindStringSubmatch(version)
+	if len(m) != 3 {
+		return shimVersion{}, errors.New("invalid shim version format")
+	}
+
+	major, err := strconv.ParseUint(m[1], 10, 0)
+	if err != nil {
+		return shimVersion{}, fmt.Errorf("invalid major version \"%s\"", m[1])
+	}
+	var minor uint64
+	if m[2] != "" {
+		minor, err = strconv.ParseUint(m[2], 10, 0)
+		if err != nil {
+			return shimVersion{}, fmt.Errorf("invalid minor version \"%s\"", m[2])
+		}
+	}
+	return shimVersion{Major: uint(major), Minor: uint(minor)}, nil
+}
+
+// mustParseShimVersion returns a new shimVersion from the supplied string.
+// This will panic if the supplied string is invalid. It is intended for
+// parsing a version from compiled in literals.
+func mustParseShimVersion(version string) shimVersion {
+	ver, err := parseShimVersion(version)
+	if err != nil {
+		panic(err)
+	}
+	return ver
+}
+
+func parseShimVersionDataIdent(r io.Reader) (shimVersion, error) {
+	scanner := bufio.NewScanner(r)
+
+	if !scanner.Scan() {
+		return shimVersion{}, errors.New("empty .data.ident section")
+	}
+	if scanner.Text() != "UEFI SHIM" {
+		return shimVersion{}, errors.New("unexpected .data.ident section contents (not shim?)")
+	}
+	for scanner.Scan() {
+		m := shimIdentVersionRE.FindStringSubmatch(scanner.Text())
+		if len(m) == 2 {
+			return parseShimVersion(m[1])
+		}
+	}
+	if scanner.Err() != nil {
+		return shimVersion{}, xerrors.Errorf("cannot decode .data.ident section contents: %w", scanner.Err())
+	}
+	return shimVersion{}, errors.New("cannot determine version - missing from .data.ident section")
+}
+
+// Compare compares 2 shim versions. It returns 0 if a == b, -1 if a < b and 1 if a > b
+func (a shimVersion) Compare(b shimVersion) int {
+	switch {
+	case a.Major > b.Major:
+		return 1
+	case a.Major < b.Major:
+		return -1
+	case a.Minor < b.Minor:
+		return -1
+	case a.Minor > b.Minor:
+		return 1
+	default:
+		return 0
+	}
+}
+
+// shimVendorCertFormat describes the format of the content of shim's .vendor_cert
+// section. This is important because it affects the format of measurements in some
+// circumstances.
+type shimVendorCertFormat int
+
+const (
+	// shimVendorCertIsX509 indicates that shim's .vendor_cert section contains
+	// a single X.509 certificate.
+	shimVendorCertIsX509 shimVendorCertFormat = iota + 1
+
+	// shimVendorCertIsDb indicates that shim's .vendor_cert section contains
+	// a signature database consisting of one or more ESLs.
+	shimVendorCertIsDb
+)
+
+// shimSbatLevel corresponds to the SBAT revocation level payload stored inside
+// shim.
+type shimSbatLevel [2][]byte
+
+// ForPolicy returns the SBAT revocation level for the specified policy.
+func (l shimSbatLevel) ForPolicy(policy shimSbatPolicy) []byte {
+	return l[policy-1]
```

Comment:
should this produce a more meaningful panic if policy is not 1 (latest) or 2  (previous) ?

Reply:
I've added a check here so that it panics with a useful message.